### PR TITLE
Remove pipe factory, immutable pipe options

### PR DESCRIPTION
**Pull request by pakrym**
Remove Unowned*
Merge extensions classes

### CODE REVIEW

File: samples/System.IO.Pipelines.Samples/CompressionSample.cs [hunk 1]
```diff
@@ -13,7 +14,7 @@ public class CompressionSample : ISample
     {
         public Task Run()
         {
-            using (var factory = new PipeFactory())
+            using (var factory = new MemoryPool())
```
Comment (ahsonkhan) [inline samples/System.IO.Pipelines.Samples/CompressionSample.cs]: nit: rename factory to pool like elsewhere in this PR.

File: src/System.IO.Pipelines.Networking.Windows.RIO/Internal/RioThread.cs [hunk 2]
```diff
@@ -27,7 +27,7 @@ internal unsafe class RioThread
         private readonly Queue<NotifyBatch> _notifyBatches;
         private readonly Queue<NotifyBatch> _processedBatches;
 
-        private PipeFactory _factory;
+        private MemoryPool _factory;
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines.Networking.Windows.RIO/Internal/RioThread.cs]: nit: rename _factory to _pool

File: samples/System.IO.Pipelines.Samples/SampleBase/RawHttpClientSampleBase.cs [hunk 3]
```diff
@@ -14,7 +15,7 @@ public abstract class RawHttpClientSampleBase : ISample
     {
         public async Task Run()
         {
-            var consoleOutput = GetPipeFactory().CreateWriter(Console.OpenStandardOutput());
+            var consoleOutput = StreamPipeConnection.CreateWriter(new PipeOptions(GetBufferPool()), Console.OpenStandardOutput());
```
Comment (davidfowl) [inline samples/System.IO.Pipelines.Samples/SampleBase/RawHttpClientSampleBase.cs]: Messy

File: samples/System.IO.Pipelines.Samples/Socket/RawSocketHttpClientSample.cs [hunk 4]
```diff
@@ -10,23 +11,23 @@ namespace System.IO.Pipelines.Samples
...
-            factory = new PipeFactory();
+            pool = new MemoryPool();
         }
 
         protected override Task<IPipeConnection> GetConnection()
         {
             Socket s = new Socket(SocketType.Stream, ProtocolType.Tcp);
             s.Connect(new IPEndPoint(IPAddress.Loopback, 5000));
-            return Task.FromResult(factory.CreateConnection(new NetworkStream(s)));
+            return Task.FromResult((IPipeConnection)new StreamPipeConnection(new PipeOptions(pool), new NetworkStream(s)));
```
Comment (davidfowl) [inline samples/System.IO.Pipelines.Samples/Socket/RawSocketHttpClientSample.cs]: Some local variables can clean this up.

File: samples/System.IO.Pipelines.Samples/SampleBase/RawHttpClientSampleBase.cs (hunk 3 above)
Comment (KrzysztofCwalina) [reply samples/System.IO.Pipelines.Samples/SampleBase/RawHttpClientSampleBase.cs]: Could we default to some pool instead of having to always pass it in?

Comment (pakrym) [reply samples/System.IO.Pipelines.Samples/SampleBase/RawHttpClientSampleBase.cs]: It has to be disposed, making options `IDisposable` would be a bit strange.

**Pull request merged**